bridge/jira/export.go

```diff
@@ -0,0 +1,449 @@
+				out <- core.NewExportError(errors.Wrap(err, "can't load bug"), id)
+				return
+			}
+
+			select {
+
+			case <-ctx.Done():
+				// stop iterating if context cancel function is called
+				return
+
+			default:
+				snapshot := b.Snapshot()
+
+				// ignore issues whose last modification date is before the query date
+				// TODO: compare the Lamport time instead of using the unix time
+				if snapshot.CreatedAt.Before(since) {
+					out <- core.NewExportNothing(b.Id(), "bug created before the since date")
+					continue
+				}
+
+				if snapshot.HasAnyActor(allIdentitiesIds...) {
+					// try to export the bug and it associated events
+					self.exportBug(ctx, b, since, out)
+				} else {
+					out <- core.NewExportNothing(id, "not an actor")
+				}
+			}
+		}
+	}()
+
+	return out, nil
+}
+
+// exportBug publish bugs and related events
+func (self *jiraExporter) exportBug(
+	ctx context.Context, b *cache.BugCache, since time.Time,
+	out chan<- core.ExportResult) {
+	snapshot := b.Snapshot()
+
+	var bugJiraID string
+
+	// Special case:
+	// if a user try to export a bug that is not already exported to jira (or
+	// imported from jira) and we do not have the token of the bug author,
+	// there is nothing we can do.
+
+	// first operation is always createOp
+	createOp := snapshot.Operations[0].(*bug.CreateOperation)
+	author := snapshot.Author
+
+	// skip bug if it was imported from some other bug system
+	origin, ok := snapshot.GetCreateMetadata(keyOrigin)
+	if ok && origin != target {
+		out <- core.NewExportNothing(
+			b.Id(), fmt.Sprintf("issue tagged with origin: %s", origin))
+		return
+	}
+
+	// skip bug if it is a jira bug but is associated with another project
+	// (one bridge per JIRA project)
+	project, ok := snapshot.GetCreateMetadata(keyJiraProject)
+	if ok && !stringInSlice(project, []string{self.project.ID, self.project.Key}) {
+		out <- core.NewExportNothing(
+			b.Id(), fmt.Sprintf("issue tagged with project: %s", project))
+		return
+	}
+
+	// get jira bug ID
+	jiraID, ok := snapshot.GetCreateMetadata(keyJiraID)
+	if ok {
+		out <- core.NewExportNothing(b.Id(), "bug creation already exported")
+		// will be used to mark operation related to a bug as exported
+		bugJiraID = jiraID
+	} else {
+		// check that we have credentials for operation author
+		client, err := self.getIdentityClient(&ctx, author.Id())
+		if err != nil {
+			// if bug is not yet exported and we do not have the author's credentials
+			// then there is nothing we can do, so just skip this bug
+			out <- core.NewExportNothing(
+				b.Id(), fmt.Sprintf("missing author token for user %.8s",
+					author.Id().String()))
+			return
+		}
+
+		// Load any custom fields required to create an issue from the git
+		// config file.
+		fields := make(map[string]interface{})
+		defaultFields, hasConf := self.conf[keyCreateDefaults]
+		if hasConf {
+			err = json.Unmarshal([]byte(defaultFields), &fields)
+			if err != nil {
+				panic("Invalid JSON in config")
```

Comment:
in go, the code should pretty much never panic. Just return an error.

Reply:
You are correct that nothing terrible can happen with this one, but it's also a question of UX. A stacktrace is not exactly user friendly ;)

Reply:
Done.